File: ios/Classes/registry/WEPropertyRegistry.swift
import WEPersonalization

class WEPropertyRegistry{
    private init(){}
    
    static let shared = WEPropertyRegistry()
    
    private var registryMap = [Int:WEProperty]()
    internal var impressionTrackedForTargetviews:[String] = []
    
    
    public func registerProperty(details:[String:Any]){
        WELogger.d("WEP I : register property for \(details)")
        let id = details[WEConstants.PAYLOAD_ID] as! Int
        let weghinline = WEProperty(id: id,
                                    screenName: details[WEConstants.PAYLOAD_SCREEN_NAME] as! String,
                                    propertyID: details[WEConstants.PAYLOAD_IOS_PROPERTY_ID] as! Int)
        
        registryMap[id] = weghinline
        WEPersonalization.shared.registerWEPlaceholderCallback(weghinline.propertyID, self)
        
    }
    
    public func deRegisterProperty(details:[String:Any])->Bool{
        WELogger.d("WEP I : deregister property for \(details)")
        let id = details[WEConstants.PAYLOAD_ID] as! Int
        return deRegisterProperty(forId: id)
    }
    
    private func deRegisterProperty(forId:Int)->Bool{
        let contains = registryMap[forId] != nil
        if(contains){
            WEPersonalization.shared.unregisterWEPlaceholderCallback(registryMap[forId]!.propertyID)
        }
        registryMap.removeValue(forKey: forId)
        return contains
        
    }
    
    private func getProperty(targetViewTag:Int)->WEProperty?{
        for(_,weginline) in registryMap{
            if(weginline.propertyID == targetViewTag){
                return weginline
            }
        }
        return nil
    }
    
    //MARK: - Impression tracking details helper
    internal func setImpressionTrackedDetails(forTag:Int,campaignId:String){
        let key = "\(forTag)_\(campaignId)"
        self.impressionTrackedForTargetviews.append(key)
    }
    
    internal func isImpressionAlreadyTracked(forTag:Int,campaignId:String)->Bool{
        var shouldReturn = false
        //        self.serialQueue.sync {
        let key = "\(forTag)_\(campaignId)"
        if self.impressionTrackedForTargetviews.first(where: {$0 == key}) != nil{
            shouldReturn = true
        }else if let trackedCampaign = self.impressionTrackedForTargetviews.first(where: { $0.hasPrefix("\(forTag)_") }) {
            // Over here we can find same property has the impression tracked for another campaign
            // so remove that tracked entry and return as false as we haven't tracked the new campaign
            // new campaign is diff to track
            self.impressionTrackedForTargetviews.removeAll(where: {$0 == trackedCampaign})
            shouldReturn = false
        }
        //        }
        return shouldReturn
    }
    
    func screenNavigated(screenName:String){
        for(_,weginline) in registryMap{
            if(weginline.screenName == screenName){
                WEPersonalization.shared.registerWEPlaceholderCallback(weginline.propertyID, self)
            }
        }
    }
    
    //MARK: - Impression tracking helper
    func trackImpression(id:Int, attributes: [String : Any]?){
        
        if(registryMap[id] != nil){
            var data = registryMap[id]!;
            WELogger.d("WEP I : track Impression for ID : \(data.propertyID) \(String(describing: attributes))")
            data.campaignData?.trackImpression(attributes: attributes)
        }
    }
    
    //MARK: - Impression tracking helper
    func trackClick(id:Int, attributes: [String : Any]?){
        if(registryMap[id] != nil){
            var data = registryMap[id]!;
            WELogger.d("WEP I : track Click for ID : \(data.propertyID) \(String(describing: attributes))")
            data.campaignData?.trackClick(actionDetails: (nil,nil), attributes: attributes)
        }
    }
    
    
}

extension WEPropertyRegistry : WEPlaceholderCallback{
    internal func onDataReceived(_ data: WECampaignData) {
        var weProperty = getProperty(targetViewTag: data.targetViewTag)
        weProperty?.campaignData = data
        WELogger.d("WEP I : onDataReceived \(weProperty != nil) \(WEPersonalizationPlugin.methodChannel != nil)")
        if(weProperty != nil){
            WEPersonalizationPlugin.methodChannel?.sendCallbacks(methodName: WEConstants.METHOD_NAME_ON_DATA_RECEIVED,
                                                                 message: WEUtils.generateMap(weginline: weProperty!,
                                                                                            campaignData: data))
        }
    }
    
    internal func onRendered(data: WECampaignData) {
        WELogger.d("WEP I : onRendered \(data.targetViewTag)")
        var weProperty = getProperty(targetViewTag: data.targetViewTag)
        weProperty?.campaignData = data
        if(weProperty != nil){
            WEPersonalizationPlugin.methodChannel?.sendCallbacks(methodName: WEConstants.METHOD_NAME_ON_RENDERED,
                                                                 message: WEUtils.generateMap(weginline: weProperty!,
                                                                                            campaignData: data))
        }
    }
    
    internal func onPlaceholderException(_ campaignId: String?, _ targetViewId: String, _ exception: Error) {
        WELogger.d("WEP I : onPlaceholderException \(String(describing: campaignId)) pId : \(targetViewId) error \(exception.localizedDescription)")
        let weProperty = getProperty(targetViewTag: Int(targetViewId) ?? -1)
        if(weProperty != nil){
            WEPersonalizationPlugin.methodChannel?.sendCallbacks(methodName: WEConstants.METHOD_NAME_ON_PLACEHOLDER_EXCEPTION,
                                                                 message: WEUtils.generateMap(weginline: weProperty!,
                                                                                            campaignId: campaignId,
                                                                                            targetViewId: targetViewId,
                                                                                            error: exception))
        }
    }
}

File: android/src/main/kotlin/com/webengage/we_personalization_flutter/registry/WEPropertyRegistry.kt
package com.webengage.we_personalization_flutter.registry

import com.webengage.we_personalization_flutter.WEPersonalizationPlugin
import com.webengage.we_personalization_flutter.model.WEProperty
import com.webengage.we_personalization_flutter.utils.*
import com.webengage.personalization.WEPersonalization
import com.webengage.personalization.callbacks.WEPlaceholderCallback
import com.webengage.personalization.data.WECampaignData
import java.lang.Exception
import java.util.logging.Logger

internal class WEPropertyRegistry {

    companion object {
        val instance = WEPropertyRegistry()
    }

    private var registryMap: HashMap<Int, WEProperty> = hashMapOf()
    private var WEPersonalizationPlugin: WEPersonalizationPlugin? = null
    private val impressionTrackedForTargetViews: MutableList<String> = mutableListOf()

    fun initFlutterPlugin(WEPersonalizationPlugin: WEPersonalizationPlugin?) {
        this.WEPersonalizationPlugin = WEPersonalizationPlugin
    }

    fun registerProperty(details: HashMap<String, Any>): Boolean {
        WELogger.v("WEPropertyRegistry","registerProperty for details $details")
        val id = details[PAYLOAD_ID] as Int
        val weProperty = WEProperty(
            id = id,
            screenName = details[PAYLOAD_SCREEN_NAME] as String,
            propertyID = details[PAYLOAD_ANDROID_PROPERTY_ID] as String
        )
        registryMap[id] = weProperty
        load(weProperty)
        return true
    }

    private fun load(weProperty: WEProperty) {
        WEPersonalization.get()
            .registerWEPlaceholderCallback(weProperty.propertyID, object : WEPlaceholderCallback {
                override fun onDataReceived(data: WECampaignData) {
                    onDataReceived(data, weProperty)

                }

                override fun onPlaceholderException(
                    campaignId: String?,
                    targetViewId: String,
                    error: Exception
                ) {
                    onPlaceholderException(campaignId, targetViewId, error, weProperty)
                }

                override fun onRendered(data: WECampaignData) {
                    onRendered(data, weProperty)
                }

            })
    }

    fun deregisterProperty(details: HashMap<String, Any>): Boolean {
        WELogger.v("WEPropertyRegistry","deregisterProperty for details $details")
        val id = details[PAYLOAD_ID] as Int
        return deregisterProperty(id)
    }

    fun trackClick(id: Int, data: HashMap<String, Any>) {
        WELogger.v("WEPropertyRegistry","trackClick called for $data")
        if (registryMap.containsKey(id)) {
            val inline = registryMap[id];
            inline?.weCampaignData?.trackClick(data)
        }
    }

    fun trackImpression(id: Int, data: HashMap<String, Any>) {
        WELogger.v("WEPropertyRegistry","trackImpression called for $data")
        if (registryMap.containsKey(id)) {
            val inline = registryMap[id];
            inline?.weCampaignData?.trackImpression(data)
        }
    }

    private fun deregisterProperty(id: Int): Boolean {
        val contains = registryMap.containsKey(id)
        if (contains) {
            val data = registryMap[id]
            WEPersonalization.get().unregisterWEPlaceholderCallback(data!!.propertyID)
        }
        registryMap.remove(id)

        return contains
    }

    private fun onDataReceived(data: WECampaignData, weProperty: WEProperty) {
        WELogger.v("WEPropertyRegistry","onDataReceived ${data.targetViewId}")
        weProperty.weCampaignData = data
        WEPersonalizationPlugin?.sendCallback(
            METHOD_NAME_ON_DATA_RECEIVED,
            WEUtils.generateMap(weProperty, data)
        )
    }

    private fun onPlaceholderException(
        campaignId: String?,
        targetViewId: String,
        error: Exception,
        weProperty: WEProperty
    ) {
        WELogger.v("WEPropertyRegistry","onPlaceholderException")
        WEPersonalizationPlugin?.sendCallback(
            METHOD_NAME_ON_PLACEHOLDER_EXCEPTION,
            WEUtils.generateMap(weProperty, campaignId, targetViewId, error)
        )
    }

    private fun onRendered(data: WECampaignData, weProperty: WEProperty) {
        WELogger.v("WEPropertyRegistry","onRendered")
        weProperty.weCampaignData = data
        WEPersonalizationPlugin?.sendCallback(
            METHOD_NAME_ON_RENDERED,
            WEUtils.generateMap(weProperty, data)
        )
    }

    fun setImpressionTrackedDetails(targetViewId: String, campaignId: String) {
        impressionTrackedForTargetViews.add("${targetViewId}_${campaignId}")
    }

    fun isImpressionAlreadyTracked(targetViewId: String, campaignId: String): Boolean {
        return impressionTrackedForTargetViews.contains("${targetViewId}_${campaignId}")
    }

    fun clearCacheData() {
        impressionTrackedForTargetViews.clear()
    }

    fun onScreenNavigated(screenName: String) {
        for ((key, value) in registryMap) {
            if (value.screenName.isNotEmpty() && screenName == value.screenName) {
                value.weCampaignData = null
                load(value)
            }
        }
    }

}